### PR TITLE
fix(ui): virtualize heavy chat history

### PR DESCRIPTION
**Pull request by karimodm**
## Summary

- virtualize tool-heavy chat history based on nested message count instead of turn count alone
- keep only the actively changing final turn outside the history virtualizer
- move completed retry/busy tails into virtualized history while preserving user-only waiting states
- preserve trailing system/ungrouped message order after the live turn
- reduce overscan for heavy turns and estimate row heights from nested message counts plus bounded measured samples
- preserve the current scroll offset when a desktop history transitions into virtualization
- cover live-turn ownership, parent-linked edge cases, tail ordering, heavy-entry activation, overscan, and entry-size estimates with focused tests

## Why

Long, tool-heavy sessions can contain dozens of assistant messages inside only a few turns. The existing five-turn threshold and eight-row overscan therefore leave most or all message/tool DOM mounted. The final turn also remains outside the virtualizer after streaming finishes.

This change makes windowing account for the actual cost of each turn while retaining the existing behavior for small desktop conversations and mobile history.

## Performance

Representative production profile of the reported long session:

| Metric | Before | After initial load | Change |
| --- | ---: | ---: | ---: |
| Mounted messages | 145 | 23 | -84% |
| DOM nodes | 18,810 | 9,527 | -49% |
| JS heap | ~288 MB | ~178 MB | -38% |
| Longest main-thread task | 1.17 s | 866 ms | -26% |

The DOM and memory reductions are substantial, but the measured timing improvement is modest and an 866 ms task remains too long. First useful message rendering was approximately 664 ms. Keystroke-to-paint and session-switch latency were not measured, so this PR does not claim that those interactions are fully fixed.

Older history remains paginated and loaded when the user reaches the top. Browser verification also confirmed normal upward scrolling and an exact 0 px bottom gap after returning to the latest message.

## Limitation and next step

Virtualization remains turn-level. A single exceptionally large visible turn can still mount more than 100 messages and dominate main-thread work. Nested assistant-message virtualization was evaluated but omitted because it regressed prepend anchoring and message-target navigation.

A follow-up should add message/tool-level progressive rendering or nested virtualization while preserving:

- prepend anchor stability while older history loads
- `scrollToMessageId` navigation to currently unmounted messages
- streaming-tail behavior and bottom pinning
- keyboard and selection behavior across virtualized boundaries

That follow-up should benchmark keystroke-to-paint latency, session-switch latency, total blocking time, and longest tasks rather than relying only on DOM and heap reductions.

## Validation

- `bun run type-check`
- `bun run lint`
- 25 focused tests across live-turn ownership, parent-linked edge cases, tail ordering, history windowing, timeline anchoring, turn windows, and streaming-tail behavior
- `bun run build:web`
- production browser checks for initial load, older-history pagination, upward scrolling, and bottom pinning

Addresses #2267
Related to #2549

### CODE REVIEW

File: packages/ui/src/components/chat/MessageList.tsx [hunk 1]
```diff
@@ -1085,10 +1106,20 @@ const StaticHistoryList = React.memo(({ entries, engine, contentRef, scrollRef,
         const sizes = tanstackVirtualizer.itemSizeCache;
         if (sizes.size >= TANSTACK_ESTIMATE_MIN_SAMPLES) {
             let total = 0;
-            for (const size of sizes.values()) total += size;
-            estimatedEntrySizeRef.current = Math.min(
+            let messageCount = 0;
+            let sampleCount = 0;
+            for (const [key, size] of sizes) {
+                const count = entryMessageCounts.get(String(key));
+                if (count === undefined) continue;
```
[the remark this answers is not in the document]
Comment (karimodm) [reply packages/ui/src/components/chat/MessageList.tsx:1120]: Checked against the installed `@tanstack/virtual-core` 3.17.3 implementation: `itemSizeCache` is populated with `item.key`, and `item.key` comes from the configured `getItemKey(index)`. This component configures `getItemKey` to return the stable render-entry ID, so the cache keys are `turn:...` / `msg:...`, not row indexes. `entryMessageCounts.get(String(key))` therefore uses the same key domain and the adaptive estimate does update. No code change is needed for this thread.

File: packages/ui/src/components/chat/MessageList.tsx [hunk 2]
```diff
@@ -1378,15 +1409,31 @@ const MessageList = React.forwardRef<MessageListHandle, MessageListProps>(({
     }), [baseDisplayMessages, retryOverlay]);
 
     const planModeEnabled = useFeatureFlagsStore((state) => state.planModeEnabled);
+    const keepLastTurnLive = sessionIsWorking || activeStreamingMessageId !== null;
```
[the remark this answers is not in the document]
Comment (karimodm) [reply packages/ui/src/components/chat/MessageList.tsx]: Addressed in f325da25. The final turn now stays outside virtualized history only when it owns an active stream or when the busy final projected turn has no attached assistant response yet. Completed retry/busy tails enter static virtualized history. The decision is based on projected parent relationships, so synthetic, orphaned, or late assistants do not incorrectly complete the final turn. Added focused coverage for active streaming, user-only busy state, completed retry tails, earlier-parent replies, orphan assistants, and ordered trailing ungrouped messages.

File: packages/ui/src/components/chat/MessageList.tsx [hunk 3]
```diff
@@ -1085,10 +1109,20 @@ const StaticHistoryList = React.memo(({ entries, engine, contentRef, scrollRef,
         const sizes = tanstackVirtualizer.itemSizeCache;
         if (sizes.size >= TANSTACK_ESTIMATE_MIN_SAMPLES) {
             let total = 0;
-            for (const size of sizes.values()) total += size;
-            estimatedEntrySizeRef.current = Math.min(
+            let messageCount = 0;
+            let sampleCount = 0;
+            for (const [key, size] of sizes) {
+                const count = entryMessageCounts.get(String(key));
+                if (count === undefined) continue;
```
[the remark this answers is not in the document]
Comment (karimodm) [reply packages/ui/src/components/chat/MessageList.tsx:1120]: Addressed defensively in bd3bc268. The configured cache path still uses stable entry IDs, as described above, but the estimator now supports both domains explicitly: it looks up the entry ID first and falls back to `entriesRef.current[key]` when TanStack supplies a numeric default row key. This preserves current behavior and also handles restored/default-key cache states.

**Comment by btriapitsyn** [comment]
Thank you for this — the analysis behind it was right, and heavy tool-dense histories were exactly where the old engine hurt.

The reason I'm closing it rather than asking for a rebase: main just replaced the chat timeline engine wholesale (tanstack → @legendapp/list with anchored-turn scrolling), and the architecture this PR improves no longer exists. The new list virtualizes the entire timeline including the live streaming turn — there's no separate "history virtualizer + final turn outside it" split anymore, overscan staging is handled by the library's adaptive rendering, and row heights come from its own measurement pipeline. Almost every line here targets code that was deleted, so a rebase would effectively be a rewrite against a system that already does most of what this set out to do.

If you take the new engine for a spin and still see heavy histories misbehaving (activation cost on tool-dense sessions is the spot I'd probe first), a fresh issue with a repro would be very welcome — the tuning surface now is LegendList's drawDistance/adaptive render rather than overscan. Sorry this one got overtaken mid-flight, and thanks again for digging into the hard part of the codebase.
